### ISSUE TITLE
Linter Product Plan

### ISSUE THREAD

**Issue by Boshen** [oxc-project/oxc#481]
## Product Plan

Create a linter that is

* zero-config
* only checks for correctness errors (code that is outright wrong or useless - definition from [Rust Clippy](https://github.com/rust-lang/rust-clippy)

### Justification

It is impossible for us to create a drop-in replacement of ESLint, but we can take the good parts of ESLint and make our direction toward a [Rust Clippy](https://github.com/rust-lang/rust-clippy) like linter.

Stylistics and convention rules are important as linter rules but they create too many bike-shedding configurations, so they are not going to be in the scope for Oxc Linter. People can still use ESLint for these two types of rules.

~For plugins, our current plan is to use a subset of GraphQL as the DSL for writing rules, see discussion [Linter plugin via Trustfall](https://github.com/Boshen/oxc/discussions/476)~

### Tasks

- [x] https://github.com/oxc-project/oxc/issues/636
- [ ] https://github.com/oxc-project/oxc/issues/479
- [ ] https://github.com/oxc-project/oxc/issues/503
- [ ] https://github.com/oxc-project/oxc/issues/480
- [ ] https://www.npmjs.com/package/eslint-plugin-import (recommended)
- [ ] https://github.com/oxc-project/oxc/issues/1170
- [ ] https://github.com/oxc-project/oxc/issues/493
- [ ] https://github.com/oxc-project/oxc/issues/1141
- [ ] https://github.com/oxc-project/oxc/issues/492
- [ ] https://github.com/oxc-project/oxc/issues/684
- [ ] https://github.com/oxc-project/oxc/issues/1022
- [x] https://github.com/oxc-project/oxc/issues/1191

**Comment by manzt** [comment]
Think the formatting is broken on Tasks above. I can't edit to fix myself!

**Comment by roman-petrov** [comment]
Any plans to add [eslint-plugin-perfectionist](https://github.com/azat-io/eslint-plugin-perfectionist) rules? It is becoming quite popular, 1.5 K stars. Just looking forward to fully migrate to amazing and promising OXC stack!

**Comment by Boshen** [comment]
> Any plans to add [eslint-plugin-perfectionist](https://github.com/azat-io/eslint-plugin-perfectionist) rules? It is becoming quite popular, 1.5 K stars. Just looking forward to fully migrate to amazing and promising OXC stack!

This rule fits well in the formatter but not the linter. Its documentation says "I love Prettier. However, this is not his area of responsibility." But we are an integrated toolchain, things differ a little :-)

So the answer to your question "Any plans to add [eslint-plugin-perfectionist](https://github.com/azat-io/eslint-plugin-perfectionist) rules" is currently no, but it's doable once we have more manpower and also the formatter ready.

**Comment by roman-petrov** [comment]
> So the answer to your question "Any plans to add [eslint-plugin-perfectionist](https://github.com/azat-io/eslint-plugin-perfectionist) rules" is currently no, but it's doable once we have more manpower and also the formatter ready.

Thanks! Hopefully after the OXC linter core will be ready, we will be able to use some AI tools to convert ESLint linters to OXC/Rust :)

**Comment by alxpsr** [comment]
Appreciate your efforts. Your tool is incredible. Looking forward to use it.
What about this plugins:
1. @angular-eslint
2. eslint-plugin-unused-imports
3. @nrwl/eslint-plugin-nx
4. eslint-plugin-filename-rules

Thanks in advance

**Comment by heyitsarpit** [comment]
very useful plugin for react performance - [eslint-plugin-react-refresh](https://www.npmjs.com/package/eslint-plugin-react-refresh)

**Comment by DonIsaac** [comment]
#4445

**Comment by tobiasdiez** [comment]
Graphql support would be nice as well: https://the-guild.dev/graphql/eslint/rules

**Comment by OskarLebuda** [comment]
Have you consider to add support for [eslint-plugin-airbnb](https://github.com/airbnb/javascript)? There are a lot of useful rules that a lot of project uses.

**Comment by matschik** [comment]
I would love if [eslint-plugin-svelte](https://github.com/sveltejs/eslint-plugin-svelte) got added. Svelte has to employ unique parsing for *.svelte files.

**Comment by xaddict** [comment]
oxc looks nice. I have copied over all rules from my `.eslintrc.json` file. Will there be a future addition where oxlint will log which rules aren't supported from the `rules` property? Possibly because some of them belong in the `formatter` category instead of the `linter` one?

**Comment by pumano** [comment]
@Boshen @DonIsaac I suggest to add to tasks form first post one column to table with "fixer" yes/no.

because rule may exist, but fixer not implemented for example and that case is only found problems, but not help to fix it, which eslint did. That "fixer" column help to understand support of fixers for rules

**Comment by TheElegantCoding** [comment]
I will like to see [@stylistic/eslint-plugin
](github.com/eslint-stylistic/eslint-stylistic), [eslint-plugin-tailwindcss](github.com/francoismassart/eslint-plugin-tailwindcss) And also [eslint-plugin-astro](github.com/ota-meshi/eslint-plugin-astro)

**Comment by pumano** [comment]
> What about nrwl/nx boundries plugin?

I would love to dig into it, to check how they implement it, I personally need it too)

**Comment by jozefvodicka** [comment]
Also NestJS related:
[@darraghor/eslint-plugin-nestjs-typed](https://www.npmjs.com/package/@darraghor/eslint-plugin-nestjs-typed)
[eslint-plugin-nestjs](https://www.npmjs.com/package/eslint-plugin-nestjs)

and Nuxt related
[@nuxt/eslint](https://github.com/nuxt/eslint/blob/main/packages/module) - Nuxt module generating project-aware ESLint config
[@nuxt/eslint-config](https://github.com/nuxt/eslint/blob/main/packages/eslint-config) - Static ESLint Config
[@nuxt/eslint-plugin](https://github.com/nuxt/eslint/blob/main/packages/eslint-plugin) - ESLint Plugin

**Comment by Spoutnik97** [comment]
When will oxlint considered as stable and usable in production ?
When all the eslint recommended rules of the above categories will be implemented ?

**Comment by DonIsaac** [comment]
> When will oxlint considered as stable and usable in production ?
> 
> When all the eslint recommended rules of the above categories will be implemented ?

Oxlint is usable right now.

It's subject to further breaking changes but it's quite ready for production use.

**Comment by Spoutnik97** [comment]
The website alerts : 
"

At the current stage, oxlint is not intended to fully replace ESLint; it serves as an enhancement when ESLint's slowness becomes a bottleneck in your workflow.

We recommend running oxlint before ESLint in your lint-staged or CI setup for a quicker feedback loop, considering it only takes a few seconds to run on large codebases.
"

When is it planned oxlint could replace eslint ? (Except for custom plugins, pr specific ones)

**Comment by robw-mercury** [comment]
Are there plans to support a JS config file, ideally that follows some of the ESLint flat config patterns? Our repos are using the flat config and it would be very helpful to share some config and settings between ESLint and OXLint.

**Comment by Sysix** [comment]
hey @robw-mercury 
no there is no plan at the moment to support js config files. executing js and calling it back to rust is a big overhead.
The best I can do for you is this eslint plugin: https://github.com/oxc-project/eslint-plugin-oxlint with `buildFromOxlintConfigFile`.
This disables all the rules which are configurated with oxlint config file.

**Comment by GerroDen** [comment]
Would you consider to integrate https://github.com/Rel1cx/eslint-react? They added a lot of useful rules to the set of `eslint-plugin-react`. Not all rules require types, so it would be possible.

**Comment by Sheraff** [comment]
I'm not sure how realistic/doable this is but [`eslint-plugin-react-compiler`](https://www.npmjs.com/package/eslint-plugin-react-compiler) would probably be pretty useful to a lot of people.

**Comment by onlywei** [comment]
[eslint-plugin-i18next](https://github.com/edvardchen/eslint-plugin-i18next) would be a big one for us

**Comment by Bricklou** [comment]
What about [angular-eslint](https://github.comangular-eslint/angular-eslint) or [eslint-plugin-storybook](https://github.com/storybookjs/eslint-plugin-storybook) ?

**Comment by vadymstebakov** [comment]
It would be cool to have [@tanstack/eslint-plugin-query](https://github.com/TanStack/query/tree/main/packages/eslint-plugin-query)

**Comment by chvanam** [comment]
Would be amazing if oxlint gets a native [eslint-plugin-compat](https://github.com/amilajack/eslint-plugin-compat) (browser compatibility check) based on it's own [oxc-browserslist](https://github.com/oxc-project/oxc-browserslist)!! 

If the goal is for VoidZero to be the undisputed web tooling king, this would be a must!

**Comment by leggomuhgreggo** [comment]
I'd like to submit [eslint-plugin-functional](https://github.com/eslint-functional/eslint-plugin-functional) for consideration. Thanks!

**Comment by ColtHands** [comment]
Are there any decisions on other, less popular, but still useful rules present in biome or sonar?

For example `no-useless-continue`:

1. ref 1 - https://biomejs.dev/linter/rules/no-useless-continue/
2. ref 2 - https://github.com/SonarSource/SonarJS/blob/master/packages/jsts/src/rules/S3626/rule.ts

**Comment by camc314** [comment]
`no-useless-continue` is not planned currently however if we had a contribution implementing the rule, we woulc almost certainly accept it.

Currently we are working on custom JS plugins.

**Comment by mikestopcontinues** [comment]
Are you open to adding support for `@stylistic/eslint-plugin`?

**Comment by GeoffreyParrier** [comment]
It seems that #11440 is missing in the list of tasks

**Comment by ColtHands** [comment]
Could [`eslint-plugin-json`](https://www.npmjs.com/package/eslint-plugin-json) and [`@eslint/css`](https://www.npmjs.com/package/@eslint/css), or [`eslint-plugin-css`](https://www.npmjs.com/package/eslint-plugin-css) be supported?

At very least in the plans, in theory or in far future?

**Comment by puiks** [comment]
Is there any plan to support [plugin-unused-imports](https://github.com/sweepline/eslint-plugin-unused-imports), or to implement a `no-unused-imports` rule directly?

**Comment by camc314** [comment]
> Is there any plan to support [plugin-unused-imports](https://github.com/sweepline/eslint-plugin-unused-imports), or to implement a `no-unused-imports` rule directly?

@puiks no-unused-vars should report these errors. please report an issue if not

**Comment by puiks** [comment]
eh.. There's a slight difference in my configs.` no-unused-vars` is off, while `no-unused-import` is configured independently

**Comment by camc314** [comment]
> eh.. There's a slight difference in my configs.` no-unused-vars` is off, while `no-unused-import` is configured independently

We don't have a rule for `no-unused-imports`  as it's already covered by `no-unused-vars`. I think the best solution is to enable the `no-unused-vars` rule?

**Comment by MattisAbrahamsson** [comment]
Would love a version of [import/order](https://github.com/import-js/eslint-plugin-import/issues/1665) or Biome's [organizeImports](https://biomejs.dev/assist/actions/organize-imports/), couldn't use Biome due to their VSCode extension still having some issues in monorepos and would be nice to use oxc for everything

**Comment by GerroDen** [comment]
> Would love a version of [import/order](https://github.com/import-js/eslint-plugin-import/issues/1665) or Biome's [organizeImports](https://biomejs.dev/assist/actions/organize-imports/), couldn't use Biome due to their VSCode extension still having some issues in monorepos and would be nice to use oxc for everything

we moved that in our projects to prettier with https://github.com/trivago/prettier-plugin-sort-imports.
Sadly it seems to break with the new oxcparser for prettier at the moment.

**Comment by renchap** [comment]
> Would love a version of [import/order](https://github.com/import-js/eslint-plugin-import/issues/1665) or Biome's [organizeImports](https://biomejs.dev/assist/actions/organize-imports/), couldn't use Biome due to their VSCode extension still having some issues in monorepos and would be nice to use oxc for everything

If you want a large open-source project using `import/order` with some custom rules to use as a test for the implementation, you can have a look at [Mastodon's ESLint config](https://github.com/mastodon/mastodon/blob/a485f97d21c3fb7da927e8fd239b301a590ee9b8/eslint.config.mjs#L94-L142).
I would be very happy to switch to `oxclint` once there is support for `import/order` and the type-aware rules

**Comment by ColtHands** [comment]
Some of rules what are suggested here from other users do contain stylistic/prettier suggestions. However those are "out of scope of this project".

- Can we have a decision on how to implement those stylistic rules (max_len for example)?
- Committing rules to this repository is encouraged, but stylistic rules are out of the scope?
- Should/could they be introduced with a `plugin` integration (similar to what Prettier with Intermediate Representation does)?

Basically I'm duplicating question from [here](https://github.com/oxc-project/oxc/issues/479#issuecomment-2722085997), but at least a high level clear decision on what we should expect 🙏

**Comment by OskarLebuda** [comment]
In ESLint there’s a [`quotes`](https://eslint.org/docs/latest/rules/quotes) rule. Does `oxlint` have something like that too, or is it missing?

**Comment by camc314** [comment]
> In ESLint there’s a [`quotes`](https://eslint.org/docs/latest/rules/quotes) rule. Does `oxlint` have something like that too, or is it missing?

Oxlint doesn't have this rule, and it isn't planned to be implemented. It will be implemented as part of the formatter.

**Comment by OskarLebuda** [comment]
> > In ESLint there’s a [`quotes`](https://eslint.org/docs/latest/rules/quotes) rule. Does `oxlint` have something like that too, or is it missing?
> 
> Oxlint doesn't have this rule, and it isn't planned to be implemented. It will be implemented as part of the formatter.

Do you have any roadmap for the formatter?

**Comment by camc314** [comment]
> > > In ESLint there’s a [`quotes`](https://eslint.org/docs/latest/rules/quotes) rule. Does `oxlint` have something like that too, or is it missing?
> > 
> > 
> > Oxlint doesn't have this rule, and it isn't planned to be implemented. It will be implemented as part of the formatter.
> 
> Do you have any roadmap for the formatter?

 - #10573 
 - https://github.com/oxc-project/oxc/milestone/14 
 -  https://github.com/oxc-project/oxc/issues/10179

**Comment by TheAlexLichter** [comment]
There is now [an RFC for the formatter](https://github.com/oxc-project/oxc/discussions/13608)!

**Comment by rafael-lua** [comment]
`Solid`
I would like to suggest https://github.com/solidjs-community/eslint-plugin-solid to be included in the list if possible.

**Comment by TheAlexLichter** [comment]
@rafael-lua We are working [on support for custom JS plugins](https://github.com/oxc-project/oxc/issues/9905) (with an ESLint compatible API). Ideally this also can cover the solid plugin (while still having the option to port rules to native if they will have significantly better perf or someone is up for a contribution!)

**Comment by colin-harrison** [comment]
Are there any plans for adding [eslint-plugin-testing-library](https://github.com/testing-library/eslint-plugin-testing-library) and [eslint-plugin-playwright](https://github.com/mskelton/eslint-plugin-playwright)?

Would contributions for rules from those plugins be accepted despite not being on the product plan above?

**Comment by connorshea** [comment]
> Are there any plans for adding [eslint-plugin-testing-library](https://github.com/testing-library/eslint-plugin-testing-library) and [eslint-plugin-playwright](https://github.com/mskelton/eslint-plugin-playwright)?
> 
> Would contributions for rules from those plugins be accepted despite not being on the product plan above?

These will hopefully be supported via the JS Plugins functionality soon, although currently they both error since it's not fully implemented yet. I noted the testing-library errors in the JS Plugins thread: https://github.com/oxc-project/oxc/discussions/14862

EDIT: Not sure about the playwright plugin, but I have a production codebase with eslint-plugin-testing-library that works fine with the 1.33.0 oxlint release :)

**Comment by theoephraim** [comment]
What is the long-term plan for supporting more _stylistic_ kind of rules. Will oxfmt get a ton more options? Or a full plugin system? Would love to migrate over eventually, but I still want finer control.

(I didn't see a similar issue for the formatter - so forgive me if this is the wrong place)

**Comment by camc314** [comment]
> What is the long-term plan for supporting more stylistic kind of rules. Will oxfmt get a ton more options? Or a full plugin system? Would love to migrate over eventually, but I still want finer control.

I believe it will be a mixture, oxfmt will get more options as time goes on. But it's unlikley that we will support every configuration option. Oxfmt is also planned to support prettier plugins _at some point_

We recently announced [JS Plugins Preview](https://oxc.rs/blog/2025-10-09-oxlint-js-plugins.html) for oxlint, and the [ESLint Stylistic](https://eslint.style), should just work out of the box in oxlint, and other style based eslint plugins should work as well.  Check out my post on [X](https://x.com/Cameron_C2/status/1995868081860870342?s=20) for how to use oxlint with a custom plugin (tailwind in this example).

**Comment by theoephraim** [comment]
Thanks for that info! I did just try it out and some rules seem to be working and others not. A few requests:
- how to deal with presets from js plugins? this will be a ton of rules to list out where previously we could use a preset
- diagnostics not coming through in vscode
- would like to apply `--fix` rules on save, but the oxc vscode plugin now only makes oxfmt available as a formatter. Will need to think through how we can use either just the lint rules or a combination of oxfmt and these additional rules

**Comment by connorshea** [comment]
> > Would love a version of [import/order](https://github.com/import-js/eslint-plugin-import/issues/1665) or Biome's [organizeImports](https://biomejs.dev/assist/actions/organize-imports/), couldn't use Biome due to their VSCode extension still having some issues in monorepos and would be nice to use oxc for everything
> 
> If you want a large open-source project using `import/order` with some custom rules to use as a test for the implementation, you can have a look at [Mastodon's ESLint config](https://github.com/mastodon/mastodon/blob/a485f97d21c3fb7da927e8fd239b301a590ee9b8/eslint.config.mjs#L94-L142). I would be very happy to switch to `oxclint` once there is support for `import/order` and the type-aware rules

@renchap I've actually been testing Oxlint pretty regularly on the mastodon codebase to validate the `@oxlint/migrate` tool and various React rules' behaviors :) (although the migration tool doesn't _quite_ work perfectly because the Mastodon config disables checking `js/jsx` files in a way that I'm not entirely clear on)

As for `import/order`, you should be able to accomplish most of the same change using the prettier-compatible `oxfmt` formatter now, if you'd like to explore that: https://oxc.rs/docs/guide/usage/formatter/sorting.html#sort-imports

I am unsure whether we'll implement `import/order` as it'd be extremely duplicative of the formatter now.

**Comment by 321paranoiawhy** [comment]
How about react native support? React native devs use [@react-native/eslint-plugin](https://www.npmjs.com/package/@react-native/eslint-plugin),  can it be taken into consideration?

**Comment by TheAlexLichter** [comment]
> How about react native support? React native devs use [@react-native/eslint-plugin](https://www.npmjs.com/package/@react-native/eslint-plugin), can it be taken into consideration?

did you try using this via Oxlint's JS plugins?

**Comment by 321paranoiawhy** [comment]
> > How about react native support? React native devs use [@react-native/eslint-plugin](https://www.npmjs.com/package/@react-native/eslint-plugin), can it be taken into consideration?
> 
> did you try using this via Oxlint's JS plugins?

Yes, we can try this:

```json
  "jsPlugins": ["@react-native/eslint-plugin"]
```

Hope it helps!

**Comment by yzhe554** [comment]
> Graphql support would be nice as well: https://the-guild.dev/graphql/eslint/rules

Wondering if this is on roadmap of js-plugins? I just tested and it didn't seem to work

**Comment by joris-gallot** [comment]
Hi I built [Oxlint Migration Inspector](https://github.com/joris-gallot/oxlint-migration-inspector) (a fork of [‎⁠eslint config-inspector⁠](https://github.com/eslint/config-inspector)) to help incremental ESLint → Oxlint migration
It analyzes flat ESLint configs, compares ‎⁠@oxlint/migrate⁠ scenarios (‎⁠native⁠, ‎⁠default⁠, ‎⁠max⁠), and visualizes rule coverage/gaps (‎⁠unsupported⁠, ‎⁠not_implemented⁠)
I’m using it on a monorepo with multiple configs and shared configs, hope it helps

**Comment by l0uisgrange** [comment]
Hey, why no `eslint-plugin-svelte`? Just curious.

**Comment by TheAlexLichter** [comment]
> Hey, why no `eslint-plugin-svelte`? Just curious.

This has been answered multiple times before. Svelte, Astro, Vue and other framework with **custom components** aren't easy to support because each needs their own parser. It is on the roadmap though and we are working on a performant and feasible solution.


See also our [JS Plugin Alpha Blog Post](https://oxc.rs/blog/2026-03-11-oxlint-js-plugins-alpha.html#what-it-can-t-do-yet)

**Comment by difosfor** [comment]
> So the answer to your question "Any plans to add [eslint-plugin-perfectionist](https://github.com/azat-io/eslint-plugin-perfectionist) rules" is currently no, but it's doable once we have more manpower and also the formatter ready.

That would be great! But do you mean that oxfmt will add support for the perfectionist sorting rules (not just imports) in the future? Or will it add plugin support for such a purpose somehow?

I guess for now we will just have to use the perfectionist JS plugin in oxlint while disabling the import sorting from there to leave that to oxfmt (or vice versa).

**Comment by camc314** [comment]
> > So the answer to your question "Any plans to add [eslint-plugin-perfectionist](https://github.com/azat-io/eslint-plugin-perfectionist) rules" is currently no, but it's doable once we have more manpower and also the formatter ready.
> 
> That would be great! But do you mean that oxfmt will add support for the perfectionist sorting rules (not just imports) in the future? Or will it add plugin support for such a purpose somehow?

Oxfmt has customizable import sorting, which should allow you to match perfectionist.

Prettier plugins are also planned at some point.


> I guess for now we will just have to use the perfectionist JS plugin in oxlint while disabling the import sorting from there to leave that to oxfmt (or vice versa).

Yes - this seems most reasonable at present assuming that oxfmt's import sorting doesn't meet your requirements. (if something is missing from oxfmt please open an issue)

**Comment by sushichan044** [comment]
Hello, is there a roadmap for adding native support for [`eslint-plugin-security`](https://github.com/eslint-community/eslint-plugin-security)?

I am currently running it through `jsPlugins`, which is a great temporary solution, but making them built-in rules for better performance would be awesome. Thank you for this incredible tool!

**Comment by connorshea** [comment]
> Hello, is there a roadmap for adding native support for [`eslint-plugin-security`](https://github.com/eslint-community/eslint-plugin-security)?
> 
> I am currently running it through `jsPlugins`, which is a great temporary solution, but making them built-in rules for better performance would be awesome. Thank you for this incredible tool!

For the time being there's no plan to add additional native plugins. We still a few hundred rules left to port in the ones we already support, and we're going to continue working on speeding up JS plugins over time.
